### PR TITLE
Add custom underline, beam and box cursors

### LINKED ISSUE #931: Underline cursor disappearing
I'm running alacritty on X.Org and having an interesting issue with the underline cursor. Block and Beam cursors do not show this bug.

The issue I've noticed is that when compiling Rust (or doing other things), my cursor sometimes straight-up disappears completely. Sending the codes `0`, `3` or `4` do not make the cursor appear again. Changing to a different cursor and then switching back to underline also doesn't work.

To reproduce you can just create an empty new binary Rust project and run this: `for i in {0..25}; do cargo clean && cargo run --release; done`. It seems to reliably break it for me, might require running a few more iterations sometimes tho. Another alternative is to use a project with more output lines like `alacritty`, because it seems to break when the terminal is printing to stdout while the user is not inputting text.

I've tested this before the `cursor_style` change, after the `cursor_style` change and with the `scrollback` PR's version (pretty old).

So I've just figured out something interesting. This is only an issue on one of my three screens. My DisplayPort screen works just fine, so does one of my DVI screens. But the other DVI screen will break. Here's a little showcase of what happens when moving the terminal from a working screen to the broken one and back: https://u.teknik.io/Cw5z0.webm

### PR DESCRIPTION
**Pull request by chrisduerr**
As mentioned in jwilm/alacritty#931, it can be troublesome if a font has
an underline symbol outside of the glyph's bounding box. This can lead
to the underline disappearing at the bottom of the terminal.

As a solution a symbol from the private use area was used as the
character code for the underline symbol. Whenever this symbol is
encountered, instead of rendering it, a custom block is rendered.

In this implementation the block has the full character as width and
sits flush with the bottom of the glyph's bounding box. The height is
half the distance between the baseline and the bottom of the bounding
box.

Here are some screenshots of how this underline looks now:
Small
Normal
Big
Huge

This fixes #931

### CODE REVIEW

File: font/src/ft/mod.rs [hunk 1]
```diff
@@ -294,6 +294,29 @@ impl FreeTypeRasterizer {
 
         let (pixel_width, buf) = Self::normalize_buffer(&glyph.bitmap())?;
 
+        // Render a custom symbol for the underline cursor
+        if glyph_key.c == '􊏢' {
```
Comment (jwilm) [inline font/src/ft/mod.rs]: Since we're using specific unicode area, it might be more clear to use `\u{ffff}` notation. Additionally, we should probably define a constant for this somewhere.

**Comment by NickeZ** [comment]
Could this be used to fix general underlining as well? https://github.com/jwilm/alacritty/issues/31

**Comment by chrisduerr** [comment]
I've discussed this with jwilm on IRC and while this would be possible, it's not a desired solution. As mentioned in #31, using a character for rendering underlines at all is only a quick fix and not a permanent solution. The goal would be to render underlines separately, not as special glyphs.

However what would be possible is changing the behavior of `|`. I've seen in the other issue that some people have problems with that symbol, so I could go for the same strategy as `_` and replace that with a custom rendered glyph too.

I've also looked at the darwin code and it should be fairly straight-forward to implement this for macos. So I would recommend not merging this until that is done. I'll find someone who can test any changes I make.

**Comment by maximbaz** [comment]
Oh this looks awesome on linux!

image

**Comment by chrisduerr** [comment]
I've added another commit that introduces some significant changes, so here's the commit message for visibility:

In issue jwilm/asacritty#31 a few people complained about Beam cursor
being weird with their fonts, so to make all cursors uniform in
alacritty, a custom cursor has been added for the Beam too.

The beam cursor's height is always the full height of the monospace font
glyph bounding box. The width depends on the font size. It is calculated
using the width of the cell and dividing it by two. The block is always
aligned at the far-left of the cell.

The implementation is identical to the one of the underline cursor, but
it has been refactored so the glyphs are created in `lib.rs`, which can
be used by darwin/macos to implement these features too.

A small bug in the underline cursor has also been fixed, until now the
width was just using the width of the unicode character. Now it should
be using the full width of the monospace glyph bounding box with every
font.

Here are some screenshots for the Beam cursor:
Small
Normal
Big
Huge

**Comment by maximbaz** [comment]
Oh yeah, now it is even more awesome:

image

In the right-most image the underline cursor doesn't touch `m` anymore, just like you say 🙂

I wonder if it is possible to make beam cursor also not overlap with `m`? I mean this branch is a huge improvement already, but I'm just curious 🙂

**Comment by asilvadesigns** [comment]
this looks awesome! can't wait to try

**Comment by chrisduerr** [comment]
I've added the custom cursors to macos now. There's stil a bit of testing required and tweaks to the beam cursor width will be applied. Then there's currently still an edge-case left with small font. After this everything should be good to go.

File: font/src/darwin/mod.rs [hunk 2]
```diff
@@ -454,6 +454,19 @@ impl Font {
     }
 
     pub fn get_glyph(&self, character: char, _size: f64, use_thin_strokes: bool) -> Result<RasterizedGlyph, Error> {
+        // Render custom symbols for underline and beam cursor
+        if character == super::UNDERLINE_CURSOR_CHAR {
```
Comment (jwilm) [inline font/src/darwin/mod.rs]: This will be slightly more efficient to use a match. For instance,

```rust
match character {
    UNDERLINE_CURSOR_CHAR => {
    },
    BEAM_CURSOR_CHAR => {
    },
    _ => {
        // default
    }
}
```

File: font/src/darwin/mod.rs [hunk 3]
```diff
@@ -454,6 +454,19 @@ impl Font {
...
+            let descent = -(self.ct_font.descent() as i32);
+            let width = self.glyph_advance('0') as i32;
+            return super::get_underline_cursor_glyph(descent, width);
+        } else if character == super::BEAM_CURSOR_CHAR {
+            let metrics = self.metrics();
```
Comment (jwilm) [inline font/src/darwin/mod.rs]: Might be nice to have a comment suggesting the _intent_ of these calculations in case someone needs to come work on them. Ditto for above.

File: font/src/darwin/mod.rs [hunk 4]
```diff
@@ -454,6 +454,33 @@ impl Font {
...
+            },
+            super::BEAM_CURSOR_CHAR => {
+                // Get the top of the bounding box
+                let metrics = self.metrics();
+                let height = metrics.line_height;
+                let mut ascent = height - self.ct_font.descent() + 1.;
+                if ascent.floor() == ascent {
+                    // Fix off-by-one with an exact X.0 ascent
+                    ascent -= 1.;
+                }
```
Comment (chrisduerr) [inline font/src/darwin/mod.rs]: If the ascent is exactly XXX.0, it's gonna be off by one pixel and the bar will be offset to the top. Right now I've fixed this by simply checking if it is exactly .0, however there are probably a few different approaches to this.

Adding .9999... instead of 1. would probably also fix this, but is a rather unclean solution. Would love some feedback on te approach I've taken.

**Comment by chrisduerr** [comment]
So I looked into the box cursor a bit to investigate if creating an empty box when the window is not focused would be hard to do. Apparently with the changes made in this PR, it's quite simple. So I've added this feature to this pull request, so now unfocused windows with the block cursor look like this:

Box Cursor

Another minor change I've made is instead of using half the distance between baseline and bottom of bounding box for the underline height, I'm now calculating the height of the underline based on the box width. So Beam, Box and Underline all have the same width/height.

**Comment by maximbaz** [comment]
The hollow cursor is awesome! (this fixes #794 by the way).

I'm just wondering, right now only the box cursor gets hollow, but shouldn't any cursor (i.e. also beam and underline) change to hollow box when terminal loses focus? The main point of #794 was to make it visible that the terminal is unfocused, but now this is only true for the box cursor and not for the other two.

**Comment by chrisduerr** [comment]
In theory it would be possible to make every cursor hollow, however underline and beam cursor do not really have enough "meat" to go hollow. For these kind of cursors the blinking cursor PR might help.

I haven't seen any terminals that have something like the hollow block for underline/beam, however what would be possible is to make the cursor always an empty box when the window is not focused. But that would probably need to be hidden behind a config setting, because I don't think that's something a lot of terminals do by default?

**Comment by maximbaz** [comment]
Yeah I was thinking for hollow block for all types of cursor. Maybe it's not default in other terminals, not sure, I'm just thinking that if we were designing this from scratch, "hollow block cursor means terminal is out of focus regardless of the cursor type" makes more sense to me :) 

I'll leave it to you guys to decide.

File: font/src/darwin/mod.rs (hunk 4 above)
Comment (chrisduerr) [reply font/src/darwin/mod.rs]: Turns out this is the same as using `.ceil()`, could have thought of that earlier.

**Comment by asilvadesigns** [comment]
I notice in iterm, it only hollows the cursor when it is block, not for any other.
I think this behavior is fine. As a vim user, I'll always switch back to normal mode before jumping out, but that just might be personal preference...*shrug

**Comment by maximbaz** [comment]
It's the same for me, I use vim and I don't leave it in insert mode. I'm thinking more in this direction: for vim users if beam cursor became hollow block on unfocus, we would probably not even notice and thus don't care, but for nano users (or whatever other editor would always use beam cursor) it would be important to indicate that the terminal is out of focus by changing beam cursor to a hollow block.

Also, the last merged PR was #928, which allows to change default cursor in the terminal - does it mean that if a person chooses beam cursor to be the default one, their terminal would never have hollow block on unfocus? Sounds disappointing.

**Comment by jwilm** [comment]
_urxvt_ appears to change to hollow block even when the cursor is configured to be underline. It seems reasonable to swap this out in both cases.

**Comment by chrisduerr** [comment]
I've tested this with termite, gnome-terminal and one more terminal emulator (I think it was tilix, someone else tested it), and they all did not have an empty block cursor with underline or beams.

File: font/src/lib.rs [hunk 5]
```diff
@@ -202,6 +217,75 @@ impl Default for RasterizedGlyph {
...
+        c: BEAM_CURSOR_CHAR,
+        top: ascent,
+        left: 0,
+        height,
+        width: beam_width,
+        buf: buf,
+    });
+}
+
+// Returns a custom beam cursor character
```
Comment (chrisduerr) [inline font/src/lib.rs]: There's a typo in this comment. I have this corrected locally, but I'm waiting on feedback about the box cursor for underline and beam. In case the box cursor change is not desired, this should probably still be corrected.

**Comment by unphased** [comment]
Although considering compatibility with other terms is important, I believe that if there is a behavior that makes a whole lot of sense from a UX perspective, it wouldn't hurt to make that be the new default, and also, options can be added to override it to other behaviors.

I'd like to point out that at least in iTerm, even though a beam cursor does not turn into a hollow block, when you defocus the window, it does stop blinking. 

My vote would be for the hollow block to be used in all cases for non-focused windows. This is because this feature is important to me because I use many terminals (usually 2 or 3 full-screen, running tmux).

**Comment by jwilm** [comment]
image

The box cursor doesn't quite work for me; it actually falls of the screen at the top. It seems to be about 1 pixel taller and 1 pixel narrower than the standard filled-in background.

**Comment by maximbaz** [comment]
I haven't noticed this before, but if you want to talk about pixel imperfections, here are some gifs for you 🙂

I increased font size to `60` to make the gifs, open them in a new tab to better see the imperfections.

1. Hollow block (higher, narrower):

hollow

hollow_2

2. Beam (higher):

beam

beam_2

3. Underscore (higher, narrower):

underscore

underscore_2

Hope that helps!

**Comment by chrisduerr** [comment]
I definitely want to get this pixel perfect, but just to make sure I'm changing the right stuff, is this only on linux or is macos also incorrect? Because from my tests macos should be spot-on.

**Comment by maximbaz** [comment]
I'm on Arch Linux myself, and based on the fact that there is [an output of `xprop` tool for `jwilm-desk` host](https://github.com/jwilm/alacritty/issues/348#issuecomment-299320571) posted elsewhere (which I think exists only for Linux) I'm speculating that `jwilm-desk` is a Linux box 😃 

/Sherlock 🕵️

**Comment by chrisduerr** [comment]
Smart thinking. :D 
I'll look into the linux code again then. And I'm just gonna assume macos works fine. I've checked that a lot more.

**Comment by chrisduerr** [comment]
So this should work properly on linux now. Assuming that my "hack" and the assumptions made with it are correct.

I'm not sure how reliable the approach is to use the metrics for font_key `0` instead of the actual font key, but it works on my system and if this always is the default font it might be an okay approach.

If you have any other ideas how to reliably get the bounding size, please let me know.

**Comment by maximbaz** [comment]
Just tried, it is pixel-perfect on my system too now - thanks!

**Comment by unphased** [comment]
Just noting that (on a MacBook) the HiDPI non-double resolutions seem to exhibit imperfections with this hollow box 

image

The reason is that scaling a 3360x2100 buffer down to 2880x1800 isn't going to be pixel-perfect, so it will butcher a 1-px-wide hollow box. Suggestion would be to ensure the hollow box's minimum thickness is 2 pixels wide instead of 1.

**Comment by chrisduerr** [comment]
@unphased It is necessary that the box can drop to just one pixel thickness. From what I can tell, most terminal emulators use single pixel underline, beam and box. Especially with small font sizes a 2px border width might already lead to issues. So setting a min width of 2 pixels is definitely not a solution to this problem.

**Comment by unphased** [comment]
@chrisduerr Yes, sorry I wasn't more clear. 

The way that HiDPI works in macOS/OS X is a little unintuitive, but leads to beautiful results. Basically, the "target" resolution is assembled at the double of this resolution. 

This is one article that kind of explains what's going on... https://medium.com/elementaryos/what-is-hidpi-and-why-does-it-matter-b024eabea20d

Alacritty already has an understanding of the notion of HiDPI which can be seen on macOS by spawning the app in a non-HiDPI external monitor and on a HiDPI monitor.  Let's suppose for the sake of argument that a 12pt font is rendered so that glyphs have a height of 12px on a normal screen, when going to a HiDPI screen to maintain the same size, it is 24px high. 

The min border cursor width being 2px I mean in the doubled-resolution case. So this would become a minimum 1pt wide hollow box. 

Think of it this way. on HiDPI pixel doubled screens, when the width of these things are rendered in rasterized screen space, the measurements are off by a power of two...

**Comment by chrisduerr** [comment]
Sounds to me like the screen is broken. :P 

I'm basing the width of the box borders on the width of the glyph. So if the glyph is double the width, it should also double the width of the box border. I can't really test anything on HiDPI, so if there's a fix for it that doesn't affect others, someone that is able to test it should probably look into it.

**Comment by unphased** [comment]
This particular thing I brought up is probably lowest-possible priority, admittedly. Really not a big deal.

**Comment by chrisduerr** [comment]
@unphased If you want to take a look at it and need help, feel free to ask for anything you don't understand. I'd look at it if I could but I don't think this is easily reproducable without such a screen.

Would you mind sharing an actual picture of how it looks? Just out of curiosity. Not a screenshot which doesn't really show anything.

**Comment by unphased** [comment]
https://youtu.be/YHMc5vf_pFA
